qcfractal/storage_sockets/db_queries.py

```diff
@@ -0,0 +1,193 @@
+                (
+                    select opt_id from optimization_history where torsion_id = {torsion_id}
+                )  
+                order by opt.id
+        """
+
+        return self.execute_query(sql_statement, with_keys=False)
+
+
+    def _get_final_molecules(self, torsion_id=None):
+
+        if torsion_id is None:
+            self._raise_missing_attribute('final_molecules', 'torsion drive id')
+
+        sql_statement = f"""
+                select molecule.* from molecule
+                join optimization_procedure as opt
+                on molecule.id = opt.final_molecule
+                where opt.id in
+                    (select opt_id from optimization_history where torsion_id = {torsion_id})
+        """
+
+        return self.execute_query(sql_statement, with_keys=True)
+
+    def _get_return_results(self, torsion_id=None):
+        """All return results ids of a torsion drive"""
+
+        if torsion_id is None:
+            self._raise_missing_attribute('return_results', 'torsion drive id')
+
+        sql_statement = f"""
+                select opt_res.opt_id, result.id as result_id, result.return_result from result
+                join opt_result_association as opt_res
+                on result.id = opt_res.result_id
+                where opt_res.opt_id in 
+                (
+                    select opt_id from optimization_history where torsion_id = {torsion_id}
+                )
+        """
+
+        return self.execute_query(sql_statement, with_keys=False)
+
+    def _get_all_opt_results(self, opt_ids : List[Union[int, str]]=None):
+        """Returns all the results objects (trajectory) of each optmization
+        Returns list(list) """
+
+        if opt_ids is None:
+            self._raise_missing_attribute('all_opt_results', 'List of optimizations ids')
+
+        # row_to_json(result.*)
+        sql_statement = text("""
+            select opt_id, json_agg(result.*) as trajectory_results from result
+            join opt_result_association as traj
+            on result.id = traj.result_id
+            where traj.opt_id in :opt_ids
+            group by opt_id
+        """)
+
+        # bind and expand ids list
+        sql_statement = sql_statement.bindparams(bindparam("opt_ids", expanding=True))
+
+        return self.execute_query(sql_statement, opt_ids=list(opt_ids))
+
+
+    def _get_best_opt_results(self, opt_ids : List[Union[int, str]]=None):
+        """Return the actual results objects of the best result in each optimization"""
+
+        if opt_ids is None:
+            self._raise_missing_attribute('best_opt_results', 'List of optimizations ids')
+
+        sql_statement = text("""
+            select opt_id, result.* from result
+            join (
+                select opt.opt_id, opt.result_id, max_pos from opt_result_association as opt
+                inner join (
+                        select opt_id, max(position) as max_pos from opt_result_association
```

Comment:
`best` is defined as the `position == torsiondrive.minimum_positions`. It gets worse as `torsiondrive.minimum_positions` is a dictionary of `key: position`.

I thought we had discussed dropping torsiondrives entirely as query optimization points due to this issue?

Reply:
This table we are looking up is not the torsion_drive history, this table is the trajectory of a single optimization_procedure. So, I'm assuming the REST gave me a list of opt_ids already.

Reply:
Shouldn't this be moved to an `OptimizationQuery` class in this case?